src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs

```diff
@@ -947,30 +947,64 @@ private static void ValidateTrustAttribute(TRUSTED_DOMAIN_INFORMATION_EX domainI
+                for (int i = 0; i < destination.Length; i++)
+                {
+                    destination[i] = PasswordCharacterSet[RandomNumberGenerator.GetInt32(PasswordCharacterSet.Length)];
+                }
+            });
+#else
             char[] cBuf = new char[PASSWORD_LENGTH];
+            byte[] randomBuffer = new byte[1];
 
-            using (RandomNumberGenerator RNG = RandomNumberGenerator.Create())
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
             {
-                RNG.GetBytes(buf);
-                for (int iter = 0; iter < PASSWORD_LENGTH; iter++)
+                for (int i = 0; i < PASSWORD_LENGTH; i++)
                 {
-                    int i = (int)(buf[iter] % 87);
-                    if (i < 10)
-                        cBuf[iter] = (char)('0' + i);
-                    else if (i < 36)
-                        cBuf[iter] = (char)('A' + i - 10);
-                    else if (i < 62)
-                        cBuf[iter] = (char)('a' + i - 36);
-                    else
-                        cBuf[iter] = s_punctuations[i - 62];
+                    cBuf[i] = PasswordCharacterSet[GetRandomByte(rng, PasswordCharacterSet.Length, randomBuffer)];
                 }
             }
 
-            password = new string(cBuf);
+            return new string(cBuf);
+
+            // This is a private copy of RandomNumberGenerator.GetInt32 that works
+            // on platforms that don't have it. Since this is for a private specific
+            // use, it can be simplified to assume some things about the ranges.
+            // We know the upper-bound is < 255, so we can avoid converting random
+            // bytes to integers and mask a single byte instead. We also know the
+            // lower-bound is 0.
+            // randomBuffer is a scratch buffer that is populated with random data
+            // so we don't allocate an array per-invocation.
+            static int GetRandomByte(RandomNumberGenerator rng, int toExclusive, byte[] randomBuffer)
+            {
+                Debug.Assert(toExclusive > 0 && toExclusive < byte.MaxValue);
+                Debug.Assert(randomBuffer.Length == 1);
+
+                // The total possible range is [0, 255).
+                int range = toExclusive - 1;
 
-            return password;
+                // Create a mask for only the applicable bits.
+                int mask = range;
+                mask |= mask >> 1;
+                mask |= mask >> 2;
+                mask |= mask >> 4;
+                // Don't need >> 8 or >> 16 since it is known the range is less than 255.
+
+                int result;
+
+                do
+                {
+                    rng.GetBytes(randomBuffer);
+                    result = mask & randomBuffer[0];
+                }
+                while (result > range);
+
+                Debug.Assert(result < toExclusive);
+                return result;
+            }
```

Comment:
There's no way to achieve the desired goals here without implementing a custom non-biased GetInt32 method?

Reply:
@stephentoub For .NET Standard 2.0, not to my knowledge. Based on an off-GitHub conversation with @bartonjs I think he was on the same page that we'd have to lift and adapt `GetInt32`.

Reply:
Well, we could trim the alphabet, but that speeds up brute force (especially if we trim it down to 1 :smile:).

The alternative to dropping this copy here is something like

```C#
byte[] random = new byte[SomeNumber];
char[] output = new char[Whatever];
int randomPos = random.Length;

for (int writePos = 0; writePos < output.Length; writePos++)
{
    int randomVal;

    while (true)
    {
        if (randomPos >= random.Length)
        {
            RandomNumberGenerator.GetBytes(random);
            randomPos = 0;
        }

        randomVal = random[randomPos];
        randomPos++;
 
        Debug.Assert(alphabet.Length > 64 && alphabet.Length <= 128);
        randomVal &= 127;

        if (randomVal < alphabet.Length)
        {
            break;
        }
    }

    output = alphabet[randomVal];
}
```
Which amounts to the same thing.